extensions/amp-live-list/0.1/live-list-manager.js

```diff
@@ -0,0 +1,171 @@
+    this.url_ = this.win.location.href;
+
+    /** @private @const {function(): Promise} */
+    this.work_ = this.fetchDocument_.bind(this, this.url_);
+
+    // Only start polling when doc is ready and when the viewer is visible.
+    this.whenDocReady_().then(() => {
+      // Switch out the poller interval if we can find a lower one and
+      // then make sure to stop polling if viewer is not visible.
+      this.interval_ = Math.min.apply(Math, this.intervals_);
+      this.poller_ = new Poller(this.win, this.interval_, this.work_);
+
+      if (this.viewer_.isVisible()) {
+        this.poller_.start();
+      }
+      this.setupVisibilityHandler_();
+    });
+  }
+
+  /**
+   * Makes a request to the given url for the latest document.
+   *
+   * @param {string} url
+   * @private
+   */
+  fetchDocument_(url) {
+    return xhrFor(this.win)
+        // TODO(erwinm): add update time here when possible.
+        .fetchDocument(url)
+        .then(this.getLiveLists_.bind(this));
+  }
+
+  /**
+   * Extract the `Document` instance parsed through responseType=document
+   * from the default xhr parser. This is important because we dont want
+   * to do any parsing in the UI thread.
+   *
+   * @param {!FetchResponse} res
+   * @return {!HTMLDocument}
+   */
+  getDoc(res) {
+    const doc = res.doc();
+    user.assert(doc instanceof Document,
+        'response must be an html `Document`, got %s', doc);
+    return doc;
+  }
+
+  /**
+   * Queries the document for all `amp-live-list` tags.
+   *
+   * @param {!HTMLDocument} doc
+   */
+  getLiveLists_(doc) {
+    const lists = Array.prototype.slice.call(
+        doc.getElementsByTagName('amp-live-list'));
+    lists.forEach(this.updateLiveList_.bind(this));
+  }
+
+  /**
+   * Updates the appropriate `amp-live-list` with its updates from the server.
+   *
+   * @param {!HTMLElement} liveList
+   */
+  updateLiveList_(liveList) {
+    const id = liveList.getAttribute('id');
+    user.assert(id, 'amp-live-list must have an id.');
+    if (id in this.liveLists_) {
+      this.liveLists_[id].update(liveList);
+    }
+  }
+
+  /**
+   * Register an `amp-live-list` instance for updates.
+   *
+   * @param {number} id
+   * @param {!AmpLiveList} liveList
+   */
+  register(id, liveList) {
+    const isNotRegistered = !(id in this.liveLists_);
+    if (isNotRegistered) {
+      this.liveLists_[id] = liveList;
+      this.intervals_.push(liveList.getInterval());
+    }
+  }
+
+  /**
+   * Returns a promise that is resolved when the document is ready.
+   * @return {!Promise}
+   * @private
+   */
+  whenDocReady_() {
+    return whenDocumentReady(this.win.document);
+  }
+
+  /**
+   * Listens to he viewer visibility changed event.
+   * @private
+   */
+  setupVisibilityHandler_() {
+    // Polling should always be stopped when document is no longer visible.
+    this.viewer_.onVisibilityChanged(() => {
+      if (this.viewer_.isVisible()) {
+        this.poller_.start();
```

Comment:
The design doc says to do it, so that the time since last poll is taken into account.

I think that is pretty important. Basically when you return to a page after 5 minutes you want an immediate update.